crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -63,23 +65,29 @@ impl<'db> UnionBuilder<'db> {
     ///     - TODO For enums `E` with members `X1`,...,`Xn`, replaces
     ///     `Literal[E.X1,...,E.Xn]` with `E`.
     fn simplify(&mut self) {
-        if self
-            .elements
-            .is_superset(&[Type::BooleanLiteral(true), Type::BooleanLiteral(false)].into())
-        {
-            let bool_ty = builtins_symbol_ty_by_name(self.db, "bool");
-            self.elements.remove(&Type::BooleanLiteral(true));
-            self.elements.remove(&Type::BooleanLiteral(false));
-            self.elements.insert(bool_ty);
+        if let Some(true_index) = self.elements.get_index_of(&Type::BooleanLiteral(true)) {
+            if self.elements.contains(&Type::BooleanLiteral(false)) {
+                *self.elements.get_index_mut2(true_index).unwrap() =
+                    builtins_symbol_ty_by_name(self.db, "bool");
+                self.elements.remove(&Type::BooleanLiteral(false));
+            }
         }
     }
 
     pub(crate) fn build(mut self) -> Type<'db> {
-        self.simplify();
         match self.elements.len() {
             0 => Type::Never,
             1 => self.elements[0],
-            _ => Type::Union(UnionType::new(self.db, self.elements)),
+            _ => {
+                self.simplify();
+                self.elements.shrink_to_fit();
+
+                match self.elements.len() {
+                    0 => Type::Never,
+                    1 => self.elements[0],
+                    _ => Type::Union(UnionType::new(self.db, self.elements)),
+                }
```

Comment:
We don't really need the `shrink_to_fit()` call here unless there's >=2 elements, right?

```suggestion

                match self.elements.len() {
                    0 => Type::Never,
                    1 => self.elements[0],
                    _ => {
                        self.elements.shrink_to_fit();
                        Type::Union(UnionType::new(self.db, self.elements))
                    }
                }
```